add support for encode/decode a message as an amqp data section.

amarzavery opened the pull request:
- This brings parity w.r.t sending/receiving data with amqp10 and other services in Azure.
- Fixes #60

pierreca commented on tests/misc.spec.ts:
```diff
@@ -47,6 +48,74 @@ describe("Misc tests", function () {
+    breceiver = BatchingReceiver.create((client as any)._context, partitionId, { eventPosition: EventPosition.fromEnqueuedTime(Date.now()) });
+    await client.send(obj, partitionId);
+    debug("Successfully sent the large message.");
+    const datas = await breceiver.receive(5, 5);
```
>datas [](start = 10, length = 5)

nitpick level: > 9000: data is already plural.

amarzavery replied:
🥇

princjef commented on lib/connectionContext.ts:
```diff
@@ -12,7 +13,9 @@ import { TokenProvider } from "./auth/token";
 import { SasTokenProvider } from "./auth/sas";
+import { ClientOptions } from "./eventHubClient";
 
+const isBuffer = require("is-buffer");
```
this is only really relevant for browser situations. otherwise you can just use [`Buffer.isBuffer()`](https://nodejs.org/api/all.html#buffer_class_method_buffer_isbuffer_obj)

princjef commented on lib/connectionContext.ts:
```diff
@@ -75,17 +86,54 @@ export namespace ConnectionContext {
-  export function create(config: ConnectionConfig, tokenProvider?: TokenProvider): ConnectionContext {
+  function encodeMessage(body: any): any {
+    let result: any;
+    if (body != undefined)
```
nit: i typically see this done as `body != null` rather than `body != undefined` (though i think they give you the same result)

princjef replied:
another stylistic thing: blocks like this without curly braces scare me

princjef replied:
(if you're interested: https://palantir.github.io/tslint/rules/curly/)

pierreca commented on lib/connectionContext.ts:
```diff
@@ -75,17 +86,54 @@ export namespace ConnectionContext {
+    else
+      result = rhea.message.data_section(Buffer.from("", "utf8"));
+
+    return result;
```
>result [](start = 11, length = 6)

null is going to be stringified encoded as a buffer containing 'null'. Correct?

pierreca commented on lib/connectionContext.ts:
```diff
@@ -75,17 +86,54 @@ export namespace ConnectionContext {
+    }
+    try {
+      processedBody = JSON.parse(processedBody);
+    } catch (err) {
```
>err [](start = 13, length = 3)

should you catch only SyntaxError? what other errors should you not catch?

amarzavery replied:
We should definitely log the syntax error. 

> what other errors should you not catch?

I don't see any other type of errors happening in the above lines in the function since all the statements are guarded by the if condition. However, it is worth while to have an uber try catch, log the error and return the input parameter `body` as-is.

pierreca commented on lib/eventHubClient.ts:
```diff
@@ -67,15 +91,16 @@ export class EventHubClient {
    */
-  constructor(config: ConnectionConfig, tokenProvider?: TokenProvider) {
-    this._context = ConnectionContext.create(config, tokenProvider);
+  constructor(config: ConnectionConfig, options?: ClientOptions) {
```
>options?: ClientOptions) [](start = 40, length = 24)

that's a breaking change, isn't it?

amarzavery replied:
Yes. It is. I don't expect people to use the EventHubClient constructor. They should be using `EventHubClient.createFromConnectionString("...", "..")` which should take care of instantiating the client for them.

princjef commented on lib/connectionContext.ts:
```diff
@@ -75,17 +86,54 @@ export namespace ConnectionContext {
+  function decodeMessage(body: any): any {
+    let processedBody: any = body;
+    if (isBuffer(body)) {
+      processedBody = body.toString("utf8");
```
if someone provides a buffer which is not a sequence of utf8 bytes, this will give them a weird output. the string vs. buffer case is a tough one. i'd probably just `JSON.stringify()` it even if it's a string so that you can `JSON.parse()` it on the other side and it will come out as a string again and you'll only get a buffer if you gave a buffer

amarzavery replied:
I get the idea but a little confused. Just to be sure. In the end we don't want to stringify the buffer and return it as-is if they actually sent a buffer. Otherwise JSON.parse() should take care of things.

princjef commented on lib/connectionContext.ts:
```diff
@@ -75,17 +86,54 @@ export namespace ConnectionContext {
+      processedBody = body.toString("utf8");
+    } else if (typeof body === "string") {
+      processedBody = body;
+    } else if (body.content) {
```
what's body.content?

amarzavery replied:
It is how the body comes back when we encode it as a "data_section" in rhea terms. The body is a [`Section` object](https://github.com/amqp/rhea/blob/master/lib/message.js#L74) with properties `typecode, content and multiple`. Until now, I have found the content is usually a buffer. I have seen customers doing `EventData.body.content.toString("utf8")`

princjef commented:
General comment: version bump after this should be 0.2.0

amarzavery commented on lib/eventHubClient.ts:
```diff
@@ -281,10 +306,11 @@ export class EventHubClient {
    * @returns {EventHubClient} - An instance of the eventhub client.
    */
-  static createFromConnectionString(connectionString: string, path?: string, tokenProvider?: TokenProvider): EventHubClient {
+  static createFromConnectionString(connectionString: string, path?: string, options?: ClientOptions): EventHubClient {
```
yup.

princjef replied:
the change in default encoding/decoding is sufficient to make this a breaking change. hopefully switching these to options objects will make future breaking changes less likely/frequent

amarzavery replied:
that is true.

amarzavery commented:
@princjef @pierreca - Please review the new changes.
- Addressed the review feedback. Added a DataTransformer interface with `encode` and `decode` methods.
- Added more tests.

amarzavery commented on tests/dataTransformer.spec.ts:
```diff
@@ -0,0 +1,198 @@
+  describe("decode", function () {
+    // It is possible that we receive an AMQP value type from the messages that were sent with
+    // previously shipped version of the sdk. If so then we should be able to handle those scenarios.
+    it("should correctly decode a string message body", function (done) {
```
We will send the received message body as-is if we cannot decode. This will ensure previous behavior of receiving AMQP value types remains unchanged.

princjef commented on lib/dataTransformer.ts:
```diff
@@ -0,0 +1,101 @@
+      if (isBuffer(body)) {
+        result = rhea.message.data_section(body);
+      } else if (typeof body === "string") {
+        result = rhea.message.data_section(Buffer.from(`"${body}"`, "utf8"));
```
this will work but you don't even need this case. `JSON.stringify('some text') === '"some text"'`

amarzavery replied:
True.

princjef commented on lib/dataTransformer.ts:
```diff
@@ -0,0 +1,101 @@
+      }
+    } else {
+      // convert undefined to empty string.
+      result = rhea.message.data_section(Buffer.from("\"\"", "utf8"));
```
why not just send an empty buffer? (`Buffer.alloc(0)`)

amarzavery replied:
Yes we can do that.
However while decoding we will get an error for doing JSON.parse('') (Since Buffer.alloc(0).toString() -> ''). Thus we will end up giving the empty Buffer back.

So the question is:
- Is it better to give an empty string or an empty Buffer object back when the user gave `undefined` as the message to be sent?

princjef replied:
maybe neither. i think coercing `undefined` to `null` makes more sense. that will stringify/parse correctly and null is closer to undefined than either of the other ones.

if i was picking between empty buffer and empty string, i'd pick empty buffer

amarzavery replied:
I see. I agree coercing to null is much better and will work well with other language SDKs since undefined is a JS thing.

princjef commented on lib/dataTransformer.ts:
```diff
@@ -0,0 +1,101 @@
+        debug("An error occurred while trying JSON.stringify() on the received body. " +
+          "The error is %O", err);
+        // Ensuring that the original body will be returned by resetting the orignal body back.
+        processedBody = body.content || body;
```
you don't need this. the processed body is only reassigned if the parse succeeds, at which point you can't enter this block

amarzavery replied:
true.
